Accurate overloads for `ZipFile.__init__`

max-muoto opened the pull request:
Add overload cases based on each mode to eliminate false positives for types that don't fully implement `IO[bytes]`, per https://github.com/python/typeshed/issues/10880

max-muoto commented:
Run-through of MyPy primer here:

Pip:

Removes false positive as the object they're passing in is readable, but doesn't implement `IO[bytes]`. They put a typing ignore, which is why this adds an additional warning.

werkzeug:

Removes false positive in test where a writable `ResponseStream` object is being passed into `ZipFilie.__init__`, as the mode is set to `"w"`.

sobolevn reviewed (commented):
I would consider adding test cases for this overload. It is rather complex.

max-muoto commented:
> I would consider adding test cases for this overload. It is rather complex.

Sounds good, can handle that tomorrow!

max-muoto commented:
> I would consider adding test cases for this overload. It is rather complex.

Added tests, found some minor issues which I fixed as a result: https://github.com/python/typeshed/pull/12119/commits/4cf07ab924f710fc7c656457036c7cabe93bc134

JelleZijlstra commented on stdlib/zipfile/__init__.pyi:
```diff
@@ -94,6 +94,22 @@ class ZipExtFile(io.BufferedIOBase):
     def write(self, s: str, /) -> object: ...
 
+class _ZipReadable(Protocol):
+    def seek(self, offset: int, whence: int = 0) -> int: ...
```
Protocol method arguments should usually be positional-only, unless zipfile actually calls these with keyword arguments.

max-muoto replied:
You're right, only positional arguments are used: https://github.com/python/typeshed/pull/12119/commits/97a4a951cba60d162dace2278f62b31fcf6f3f6d

max-muoto replied:
Interestingly, MyPy doesn't complain if you try and pass in an implementation that enforces positional arguments, while the protocol doesn't. Pyright seems to have the correct behavior though.

[Pyright Playground
](https://pyright-play.net/?pyrightVersion=1.1.368&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYCwAULfXY6RQIYDObUASgKYsAmLAEYUeACiJgS5CgEoAXLSjKo-HsCgg%2B-MWx4VgAGigp5mFPgC8UALQBGWbYB8UIXBg82ZgHS%2BGzdk4AQVIYAFcWCl4BYVFFGhVVdU1tXX0jEzNUK1s7YwB6RxsXNw8vJUSVLXCQFFcAInqGZpo1DWAwMDEtGJEeM2jBPqKXADkwFH6oX28Wjq6Q8MjB2PFZWVogA)

[MyPy Playground](https://mypy-play.net/?mypy=latest&python=3.12)

I'll create a MyPy issue on this.

max-muoto commented:
@JelleZijlstra Just wanted to follow up on this one!

JelleZijlstra commented:
I'll periodically look over open typeshed PRs when I have some time, but can't guarantee I'll come back to a specific one (and other maintainers may also pick up this PR if they're confident enough in the change to merge it).

max-muoto commented:
> I'll periodically look over open typeshed PRs when I have some time, but can't guarantee I'll come back to a specific one (and other maintainers may also pick up this PR if they're confident enough in the change to merge it).

Sounds good, thanks for the initial look nonetheless.

hauntsaninja commented on stdlib/zipfile/__init__.pyi:
```diff
@@ -142,6 +185,39 @@ class ZipFile:
+        @overload
+        def __init__(
+            self,
+            file: StrPath | _ZipReadable | _ZipSeekTellable,
```
Is this right? I'm not sure `_ZipSeekTellable` is enough, e.g. `_RealGetContents` requires read

max-muoto replied:
I think likely misread this, and assumed if an attribute error occurred it just needed to be tellable.

srittau commented on stdlib/zipfile/__init__.pyi:
```diff
@@ -106,24 +120,50 @@ class ZipFile:
+    # metadata_encoding is new in 3.11
     if sys.version_info >= (3, 11):
         @overload
         def __init__(
             self,
             file: StrPath | IO[bytes],
+            mode: _ZipFileMode = "r",
+            compression: int = 0,
+            allowZip64: bool = True,
+            compresslevel: int | None = None,
+            *,
+            strict_timestamps: bool = True,
+            metadata_encoding: str | None = None,
+        ) -> None: ...
```
Aren't all following overloads shadowed by this one?

max-muoto replied:
No - for example, `IO[bytes]` is broader than `_ZipReadable`, which will lead to the second overload getting chosen if you just an object that is readable.

max-muoto replied:
As for `_ZipFileMode`, if `file` doesn't match we'll end up skipping over the overload.

srittau replied:
> `IO[bytes]` is broader than `_ZipReadable`

But that's exactly the problem. It's my understanding that overloads are processed in order (although [the typing spec](https://typing.readthedocs.io/en/latest/spec/overload.html) doesn't mention that), so `IO[bytes]` always matches before `_ZipReadable` can match.

max-muoto replied:
How is that a problem though? `_ZipReadable` will get matched if `IO[bytes]` isn't fully fulfilled, or am I misunderstanding something?

hauntsaninja commented on stdlib/zipfile/__init__.pyi:
```diff
@@ -106,24 +120,50 @@ class ZipFile:
             self,
-            file: StrPath | IO[bytes],
-            mode: _ZipFileMode = "r",
+            file: StrPath | _ZipTellable | _ZipWritable,
```
While `_ZipTellable` won't immediately error, I don't think the resulting object can be used for anything right? Unless I'm missing some use case, I think we should keep it `StrPath | _ZipWritable`

hauntsaninja replied:
Oh in fact it's definitely wrong, because `__del__` will trigger `close` will trigger a write attempt

max-muoto replied:
Let me fix this.

max-muoto replied:
https://github.com/python/typeshed/pull/12119/commits/d98196ab5128bc54e5442dff57f1620eb8a67272

max-muoto replied:
@hauntsaninja Should be resolved now.

hauntsaninja reviewed (approved):
Thanks, this looks good, but had one more question!

hauntsaninja commented:
Thank you!
